torch/csrc/jit/python_compiled_function.cpp

```diff
@@ -90,18 +97,37 @@ struct CompiledFunction {
       } catch(const jit::NotImplementedException & ex) {
         closure_ = std::make_shared<AutogradClosureFactory>(complete_trace.get());
       }
+      prepareStageDetails(complete_trace.get());
+      graph_ = complete_trace->graph;
       is_ready_ = true;
       return true;
     }
 
+    void prepareStageDetails(TracingState *state) {
+      stage_details_.resize(state->graph->stage() + 1);
+      for (std::size_t stage = 0; stage < state->graph->stage() + 1; ++stage) {
+        auto & details = stage_details_[stage];
+        std::tie(details.input_flags, details.output_flags) = std::move(state->var_flags[stage]);
+        if (stage >= 1) {
+          auto & current_outputs = state->output_edges[stage];
+          auto & prev_outputs = state->output_edges[stage - 1];
+          for (auto & output : current_outputs) {
+            auto prev_it = std::find(prev_outputs.begin(), prev_outputs.end(), output);
+            if (prev_it == prev_outputs.end()) continue;
+            details.copied_next_fns.push_back(std::distance(prev_outputs.begin(), prev_it));
```

Comment:
Some comments about the indexing and layout of these arrays would be really helpful.